ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveJoinAddNotNullRule.java

```diff
@@ -74,7 +78,14 @@ public HiveJoinAddNotNullRule(Class<? extends Join> clazz,
   @Override
   public void onMatch(RelOptRuleCall call) {
     Join join = call.rel(0);
-    if (join.getJoinType() == JoinRelType.FULL || join.getCondition().isAlwaysTrue()) {
+
+    // For anti join case add the not null on right side if the condition is
```

Comment:
Not sure I understand the issue here -- is the problem the fact that ANTI-join matches with NULL rows on the right side?

Reply:
For the case when we have. join condition which gets evaluated, it will return false while comparing with a null on the right side. But for always true join condition, we will not do a match for right side assuming it's always true.  So for anti join, the left side records will not be emitted. To avoid this we put a null check on right side table and for all null entry, no records will be projected from right side and thus all records from left side will be emitted. So the comment is not very accurate. It's like even if the condition is always true, we add a null check on right side for anti join. I will update it.

Reply:
Thanks! Makes sense now